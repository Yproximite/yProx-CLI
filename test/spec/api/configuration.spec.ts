import { createFakeEnv } from '../fake-env';

const packageJson = JSON.stringify({
  yproxCli: {
    assets: {
      app: [{ handler: 'rollup', src: 'i-dont-exists/from-package-json/index.js' }],
    },
  },
});

const yproxCliConfigJs = `
module.exports = {
  assets : {
    app: [{ handler: 'rollup', src: 'i-dont-exists/from-yprox-cli-config.js/index.js'Â }]
  },
};`;

describe('api: configuration', () => {
  let processExit = process.exit;
  let consoleError = console.error;

  beforeEach(() => {
    processExit = process.exit;
    consoleError = console.error;

    // @ts-ignore
    process.exit = jest.fn();
    console.error = jest.fn();
  });

  afterEach(() => {
    process.exit = processExit;
    console.error = consoleError;
  });

  it('should load conf from `yprox-cli.config.js`', async () => {
    const { api, cleanup } = await createFakeEnv({
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
  });

  it('should load conf from `package.json`', async () => {
    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
  });

  it('should throw an error when having both config from `package.json` and `yprox-cli.config.js`', async () => {
    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(process.exit).toHaveBeenCalledWith(1);
    expect(console.error).toHaveBeenNthCalledWith(1, '[08:30:00] error :: Your configuration is invalid.');
    expect(console.error).toHaveBeenNthCalledWith(
      2,
      "[08:30:00] error :: You can't configure yprox-cli with yprox-cli.config.js and package.json at the same time."
    );

    await cleanup();
  });

  it('with YPROX_CLI_IGNORE_CONFIG_FILE=true, should load config from `package.json` even if we have two config files', async () => {
    process.env.YPROX_CLI_IGNORE_CONFIG_FILE = 'true';

    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
    delete process.env.YPROX_CLI_IGNORE_CONFIG_FILE;
  });

  it('with YPROX_CLI_IGNORE_PACKAGE_JSON_FILE=true, should load config from `yprox-cli.config.js` even if we have two config files', async () => {
    process.env.YPROX_CLI_IGNORE_PACKAGE_JSON_FILE = 'true';

    const { api, cleanup } = await createFakeEnv({
      'package.json': packageJson,
      'yprox-cli.config.js': yproxCliConfigJs,
    });

    expect(api.projectOptions).toMatchSnapshot();

    await cleanup();
    delete process.env.YPROX_CLI_IGNORE_PACKAGE_JSON_FILE;
  });
});
