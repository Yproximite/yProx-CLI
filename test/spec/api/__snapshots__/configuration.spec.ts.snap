// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api: configuration should load conf from \`package.json\` 1`] = `
Object {
  "assets": Object {
    "app": Array [
      Object {
        "dest": "dist/js",
        "destFile": "button.js",
        "handler": "rollup",
        "src": "src/components/button/index.js",
      },
      Object {
        "concat": "scripts.js",
        "dest": "dist/js",
        "handler": "js",
        "src": Array [
          "src/js/foo.js",
          "src/js/bar.js",
        ],
      },
      Object {
        "concat": "legacy-styles.css",
        "dest": "dist/css",
        "handler": "css",
        "src": Array [
          "src/css/foo.css",
          "src/css/bar.css",
        ],
      },
      Object {
        "dest": "dist/css",
        "destFile": "styles.css",
        "handler": "sass",
        "src": "src/sass/style.scss",
      },
      Object {
        "dest": "dist/images",
        "handler": "image",
        "name": "images to optimize",
        "src": "src/images/*",
      },
      Object {
        "dest": "dist/",
        "handler": "file",
        "name": "files to copy",
        "src": Array [
          "src/lorem.txt",
          "src/udhr.txt",
        ],
      },
    ],
  },
  "autoprefixer": Object {},
  "buble": Object {},
  "cssnano": Object {
    "autoprefixer": false,
    "safe": true,
  },
  "eslint": Object {
    "extensions": Array [
      ".js",
      ".vue",
    ],
  },
  "gifsicle": Object {
    "colors": 16,
    "interlaced": true,
    "optimizationLevel": 3,
  },
  "handlers": Object {
    "rollup": Object {
      "commonjs": Object {},
      "json": Object {},
      "nodeResolve": Object {},
      "shims": Object {},
      "vue": Object {},
    },
    "sass": Object {},
  },
  "jpegtran": Object {
    "progressive": true,
  },
  "optipng": Object {
    "optimizationLevel": 5,
  },
  "path": Object {},
  "svgo": Object {
    "plugins": Array [
      Object {
        "removeViewBox": true,
      },
    ],
  },
  "terser": Object {},
}
`;

exports[`api: configuration should load conf from \`yprox-cli.config.js\` 1`] = `
Object {
  "assets": Object {
    "app": Array [
      Object {
        "handler": "rollup",
        "src": "i-dont-exists/from-yprox-cli-config.js/index.js",
      },
    ],
  },
  "autoprefixer": Object {},
  "buble": Object {},
  "cssnano": Object {
    "autoprefixer": false,
    "safe": true,
  },
  "eslint": Object {
    "extensions": Array [
      ".js",
      ".vue",
    ],
  },
  "gifsicle": Object {
    "interlaced": true,
  },
  "handlers": Object {
    "rollup": Object {
      "commonjs": Object {},
      "json": Object {},
      "nodeResolve": Object {},
      "shims": Object {},
      "vue": Object {},
    },
    "sass": Object {},
  },
  "jpegtran": Object {
    "progressive": true,
  },
  "optipng": Object {
    "optimizationLevel": 5,
  },
  "path": Object {},
  "svgo": Object {
    "plugins": Array [
      Object {
        "removeViewBox": true,
      },
    ],
  },
  "terser": Object {},
}
`;

exports[`api: configuration with YPROX_CLI_IGNORE_CONFIG_FILE=true, should load config from \`package.json\` even if we have two config files 1`] = `
Object {
  "assets": Object {
    "app": Array [
      Object {
        "dest": "dist/js",
        "destFile": "button.js",
        "handler": "rollup",
        "src": "src/components/button/index.js",
      },
      Object {
        "concat": "scripts.js",
        "dest": "dist/js",
        "handler": "js",
        "src": Array [
          "src/js/foo.js",
          "src/js/bar.js",
        ],
      },
      Object {
        "concat": "legacy-styles.css",
        "dest": "dist/css",
        "handler": "css",
        "src": Array [
          "src/css/foo.css",
          "src/css/bar.css",
        ],
      },
      Object {
        "dest": "dist/css",
        "destFile": "styles.css",
        "handler": "sass",
        "src": "src/sass/style.scss",
      },
      Object {
        "dest": "dist/images",
        "handler": "image",
        "name": "images to optimize",
        "src": "src/images/*",
      },
      Object {
        "dest": "dist/",
        "handler": "file",
        "name": "files to copy",
        "src": Array [
          "src/lorem.txt",
          "src/udhr.txt",
        ],
      },
    ],
  },
  "autoprefixer": Object {},
  "buble": Object {},
  "cssnano": Object {
    "autoprefixer": false,
    "safe": true,
  },
  "eslint": Object {
    "extensions": Array [
      ".js",
      ".vue",
    ],
  },
  "gifsicle": Object {
    "colors": 16,
    "interlaced": true,
    "optimizationLevel": 3,
  },
  "handlers": Object {
    "rollup": Object {
      "commonjs": Object {},
      "json": Object {},
      "nodeResolve": Object {},
      "shims": Object {},
      "vue": Object {},
    },
    "sass": Object {},
  },
  "jpegtran": Object {
    "progressive": true,
  },
  "optipng": Object {
    "optimizationLevel": 5,
  },
  "path": Object {},
  "svgo": Object {
    "plugins": Array [
      Object {
        "removeViewBox": true,
      },
    ],
  },
  "terser": Object {},
}
`;

exports[`api: configuration with YPROX_CLI_IGNORE_PACKAGE_JSON_FILE=true, should load config from \`yprox-cli.config.js\` even if we have two config files 1`] = `
Object {
  "assets": Object {
    "app": Array [
      Object {
        "handler": "rollup",
        "src": "i-dont-exists/from-yprox-cli-config.js/index.js",
      },
    ],
  },
  "autoprefixer": Object {},
  "buble": Object {},
  "cssnano": Object {
    "autoprefixer": false,
    "safe": true,
  },
  "eslint": Object {
    "extensions": Array [
      ".js",
      ".vue",
    ],
  },
  "gifsicle": Object {
    "interlaced": true,
  },
  "handlers": Object {
    "rollup": Object {
      "commonjs": Object {},
      "json": Object {},
      "nodeResolve": Object {},
      "shims": Object {},
      "vue": Object {},
    },
    "sass": Object {},
  },
  "jpegtran": Object {
    "progressive": true,
  },
  "optipng": Object {
    "optimizationLevel": 5,
  },
  "path": Object {},
  "svgo": Object {
    "plugins": Array [
      Object {
        "removeViewBox": true,
      },
    ],
  },
  "terser": Object {},
}
`;
